fix(bazel): enable dts bundling for Ivy packages

alan-agius4 opened the pull request:
It is now possible to bundle DTS files of Ivy libraries since the blocker https://github.com/microsoft/rushstack/issues/1029 has been addressed upstream.

alan-agius4 committed:
ci: remove `@microsoft/api-extractor` from `ignoreDeps`

We can now allow Renovate to manage this dependency.

alan-agius4 committed:
fix(bazel): enable dts bundling for Ivy packages

It is now possible to bundle DTS files of Ivy libraries since the blocker https://github.com/microsoft/rushstack/issues/1029 has been addressed upstream.

alan-agius4 commented on goldens/public-api/common/common.md:
```diff
@@ -262,7 +262,7 @@ export class KeyValuePipe implements PipeTransform {
 
 // @public
-export class Location {
+class Location_2 {
```
API extractor renames these to avoid clashes with global DOM names.

petebacondarwin commented on goldens/public-api/localize/init/index.md:
```diff
@@ -5,7 +5,9 @@
 
 // @public
-export const $localize: LocalizeFn;
+const $localize_2: LocalizeFn;
+
+export { $localize_2 as $localize }
```
So why does it rename this one?

alan-agius4 replied:
I don't have an answer for that.

alan-agius4 replied:
Actually, because `$localize` is a global.
https://github.com/microsoft/rushstack/pull/1767/files

petebacondarwin reviewed (approved):
LGTM

Can you clarify the last commit message a bit? I found it a bit confusing.

Currently it says:

````
fix(compiler-cli): support reflecting local modules
With the recent change to enable Ivy DTS bundling, will convert local namespace imports to local modules such as the below;
```ts
declare namespace i1 {
  export {
    RouterOutletContract,
    RouterOutlet
  }
}

export declare class RouterModule {
    constructor(guard: any, router: Router);

    static ɵmod: i0.ɵɵNgModuleDeclaration<RouterModule, [typeof i1.RouterOutlet...]>;
}
```

With this change we amend `reflectTypeEntityToDeclaration` to support this.
````

I believe that these "local modules" are what used to be called "internal modules" and are actually now referred to simply as "namespaces". This is compared to "external modules" which are now referred to as just "modules". See https://www.typescriptlang.org/docs/handbook/namespaces.html.

I think it would be helpful to explain the background motivation for this change. I.e that when the definition of a type is in a bundled d.ts file, then what were originally namespaced imports become namespace declarations within this file. And therefore this commit adds support for reflecting types that are defined in such namespace declarations.

alan-agius4 commented:
@petebacondarwin, commit message updated.

petebacondarwin commented:
> @petebacondarwin, commit message updated.

Perfect! Thanks.

devversion reviewed (approved):
LGTM. Worth capturing here that https://github.com/angular/angular/commit/56bd21de6f62893140675c874c556fd1b95232d6 has been implemented to allow for this update to happen (due to the new alias exports being generated)

IgorMinar reviewed (approved):
LGTM! Thanks!

Reviewed-for: public-api, bazel, dev-infra

alan-agius4 commented:
Blocking until https://github.com/angular/angular/pull/42737 lands.

alan-agius4 committed:
fix(compiler-cli): support reflecting namespace declarations

DTS bundling, will cause originally namespaced imports become namespace declarations within the same file. Example:

Before bundling
```ts
import * as i1 from './router';

export declare class RouterModule {
    constructor(guard: any, router: Router);

    static ɵmod: i0.ɵɵNgModuleDeclaration<RouterModule, [typeof i1.RouterOutlet...]>;
}
```

After bundling
```
declare namespace i1 {
  export {
    RouterOutletContract,
    RouterOutlet
  }
}

export declare class RouterModule {
    constructor(guard: any, router: Router);

    static ɵmod: i0.ɵɵNgModuleDeclaration<RouterModule, [typeof i1.RouterOutlet...]>;
}
```

And therefore this commit adds support for reflecting types that are defined in such namespace declarations.

Closes #42064

jessicajaniuk reviewed (approved):
LGTM :cookie:

alan-agius4 commented:
**Caretaker**: I am unable to remove the unnecessary reviews.

atscott commented:
[presubmit](http://test/OCL:382803773:BASE:382803915:1625258071984:8df23dc7)
